File: framework/languages/java/samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/spots/CustomerEntrySpot.java
package systems.zlink.samples.deliverydispatch.server.customergateway.spots;

import systems.zlink.framework.CancellationToken;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkEntrySpot;
import systems.zlink.framework.spots.ZLinkEntrySpotContext;
import systems.zlink.framework.spots.ZLinkSpotActorJoinResponse;
import systems.zlink.samples.deliverydispatch.server.customergateway.CustomerActor;
import systems.zlink.samples.deliverydispatch.server.customergateway.CustomerActorDirectory;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

public final class CustomerEntrySpot implements ZLinkEntrySpot<CustomerActor> {
    private final ZLinkEntrySpotContext context;
    private final CustomerActorDirectory customers;

    public CustomerEntrySpot(
        ZLinkEntrySpotContext context,
        CustomerActorDirectory customers) {
        this.context = context;
        this.customers = customers;
    }

    @Override
    public ZLinkEntrySpotContext context() {
        return context;
    }

    @Override
    public void onCreateActor(
        CustomerActor actor,
        ZLinkMessage createRequest,
        CancellationToken cancellationToken) {
        customers.register(actor);
    }

    @Override
    public ZLinkSpotActorJoinResponse onActorJoin(
        String actorId,
        ZLinkMessage request,
        CancellationToken cancellationToken) {
        return ZLinkSpotActorJoinResponse.accept();
    }

    @Override
    public void onJoinedActor(
        CustomerActor actor,
        CancellationToken cancellationToken) {
        customers.register(actor);
    }

    @Override
    public void onLeaveActor(
        CustomerActor actor,
        CancellationToken cancellationToken) {
        customers.remove(actor.actorId());
    }

    public Messages.SubscribeDeliveryAccepted subscribe(
        CustomerActor actor,
        Messages.SubscribeDelivery request) {
        customers.subscribe(actor.actorId(), request.deliveryId());
        return new Messages.SubscribeDeliveryAccepted(request.deliveryId());
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/spots/CustomerEntrySpot.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.spots

import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpot
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkEntrySpotContext
import systems.zlink.framework.spots.ZLinkSpotActorJoinResponse
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.CustomerActor
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.CustomerActorDirectory
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeDeliveryReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeDeliveryRes

class CustomerEntrySpot(
    private val entryContext: ZLinkEntrySpotContext,
    private val customers: CustomerActorDirectory,
) : ZLinkSuspendingEntrySpot<CustomerActor>() {
    override fun context(): ZLinkEntrySpotContext = entryContext

    override suspend fun onCreateActorSuspending(
        actor: CustomerActor,
        createRequest: ZLinkMessage,
    ) {
        customers.register(actor)
    }

    override suspend fun onActorJoinSuspending(
        actorId: String,
        request: ZLinkMessage,
    ): ZLinkSpotActorJoinResponse = ZLinkSpotActorJoinResponse.accept()

    override suspend fun onJoinedActorSuspending(actor: CustomerActor) {
        customers.register(actor)
    }

    override suspend fun onLeaveActorSuspending(actor: CustomerActor) {
        customers.remove(actor.actorId())
    }

    fun subscribe(
        actor: CustomerActor,
        request: SubscribeDeliveryReq,
    ): SubscribeDeliveryRes {
        customers.subscribe(actor.actorId(), request.deliveryId)
        return SubscribeDeliveryRes(request.deliveryId)
    }
}
